src/emqtt.erl

```diff
@@ -828,15 +843,30 @@ mqtt_connect(State = #state{clientid    = ClientId,
                                  username     = Username,
                                  password     = emqtt_secret:unwrap(Password)}), State).
 
-reconnect(state_timeout, NextTimeout, #state{conn_mod = CMod} = State) ->
-    case do_connect(CMod, State#state{clean_start = false}) of
+reconnect(state_timeout, Attempts, #state{conn_mod = CMod} = State) ->
+    case do_connect(CMod, State) of
         {ok, #state{connect_timeout = Timeout} = NewState} ->
-            {next_state, waiting_for_connack, NewState, {state_timeout, Timeout, Timeout}};
+            {next_state, waiting_for_connack, NewState, {state_timeout, Timeout, Attempts}};
         _Err ->
-            {keep_state_and_data, {state_timeout, NextTimeout, NextTimeout*2}}
+            NextAttempts = case Attempts of
+                               infinity -> infinity;
+                               _ -> Attempts - 1
+                           end,
+            case NextAttempts =< 0 of
+                true ->
+                    {stop, reach_max_reconnect_attempts};
+                false ->
+                    #state{reconnect_timeout = ReconnectTimeout} = State,
+                    {keep_state_and_data, {state_timeout, ReconnectTimeout, NextAttempts}}
+            end
     end;
 reconnect({call, From}, stop, _State) ->
     {stop_and_reply, normal, [{reply, From, ok}]};
+reconnect(info, ?PUB_REQ(_Msg, _ExpireAt, Callback), _State) ->
+    eval_callback_handler({error, not_connected}, Callback),
```

Comment:
Why do we need this change?
see L871. The old behaviour is to postpone handling the PUB_REQ.

Reply:
The initial idea was that when integrated with `emqx_resource_buffer_worker` in EMQX. it would handle this failure and store these failed messages in a queue. _It avoids having to store a lot of messages in emqtt's queue and allows to return a failure quickly._

Reply:
> in EMQX. it would handle this failure and store these failed messages in a queue.

In that case, will that be better EMQX starts the emqtt client with reconnect = false so it get know the client process is down and restart the new one, the messages are buffered in the "buffer work" so no message will get lost.

I think If EMQTT client starts with reconnect enabled, it should buffer the Publish Request from the caller when it is reconnecting and especially it shall not fail the call (it could fail while run out of retires or delay the reply as in the old behaviour).

Fail the publish call during reconnection might confuse the caller and complicate the logic of error handling at caller side. 

WDYT?

Reply:
You are right. The original approach did make more sense. I'll change it back 👍